cpp/src/join/conditional_join.cu

```diff
@@ -13,31 +13,237 @@
+      case join_kind::LEFT_ANTI_JOIN:
+      case join_kind::FULL_JOIN: return get_trivial_left_join_indices(left, stream);
+      // Inner and left semi joins return empty output because no matches can exist.
+      case join_kind::INNER_JOIN:
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_pair(std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr),
+                              std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr));
+    }
+  }
+
+  // Prepare output column. Whether or not the output column is nullable is
+  // determined by whether any of the columns in the input table are nullable.
+  // If none of the input columns actually contain nulls, we can still use the
+  // non-nullable version of the expression evaluation code path for
+  // performance, so we capture that information as well.
+  auto const nullable  = cudf::nullable(left) || cudf::nullable(right);
+  auto const has_nulls = nullable && (cudf::has_nulls(left) || cudf::has_nulls(right));
+
+  auto const parser =
+    ast::detail::expression_parser{binary_predicate, left, right, has_nulls, stream, mr};
+  CUDF_EXPECTS(parser.output_type().id() == type_id::BOOL8,
+               "The expression must produce a boolean output.");
+
+  auto left_table  = table_device_view::create(left, stream);
+  auto right_table = table_device_view::create(right, stream);
+
+  // Allocate storage for the counter used to get the size of the join output
+  detail::grid_1d config(left_table->num_rows(), DEFAULT_JOIN_BLOCK_SIZE);
+  auto const shmem_size_per_block =
+    parser.device_expression_data.shmem_per_thread * config.num_threads_per_block;
+  join_kind KernelJoinKind = JoinKind == join_kind::FULL_JOIN ? join_kind::LEFT_JOIN : JoinKind;
+
+  // If the join size was not provided as an input, compute it here.
+  size_type join_size;
+  if (output_size.has_value()) {
+    join_size = *output_size;
+  } else {
+    rmm::device_scalar<size_type> size(0, stream, mr);
+    CHECK_CUDA(stream.value());
+    if (has_nulls) {
+      compute_conditional_join_output_size<DEFAULT_JOIN_BLOCK_SIZE, true>
+        <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
+          *left_table,
+          *right_table,
+          KernelJoinKind,
+          compare_nulls,
+          parser.device_expression_data,
+          size.data());
+    } else {
+      compute_conditional_join_output_size<DEFAULT_JOIN_BLOCK_SIZE, false>
+        <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
+          *left_table,
+          *right_table,
+          KernelJoinKind,
+          compare_nulls,
+          parser.device_expression_data,
+          size.data());
+    }
+    CHECK_CUDA(stream.value());
+    join_size = size.value(stream);
+  }
+  CUDF_EXPECTS(join_size < std::numeric_limits<size_type>::max(),
+               "The result of this join is too large for a cudf column.");
```

Comment:
Should this check be here?  The other join APIs can return a gather map that exceeds this limit, and it is up to the application to split the gather map into chunks in order to manifest the join result successfully with a subsequent `gather`.  Otherwise if we were going to enforce this limit I would expect the join APIs to return an INT32 `cudf::column` rather than a raw device buffer.

Reply:
Good point. I removed this. I also changed the size estimation to use a `std::size_t` rather than a `size_type` so that larger outputs will be supported.